docs/MultipleLoadNativeModule.md

```diff
@@ -0,0 +1,65 @@
+11. If the delegated app that was creating the service crashes, that will cause the module handle to be closed. Closing the handle will give an indication to ebpfcore to choose one of the other waiting threads (if any) to create the service entry and load the driver.
+
+#### `GET_NATIVE_MODULE_HANDLE` Design:
+1. UM will pass the service name. ebpfcore looks for the module based on the service name.
+2. If ebpfcore does not find the module, it does the following:
+    1. Generate a GUID for the client.
+    2. Create a placeholder native module and add it to hash tables.
+    3. Populate the service path (created by appending the service name to path prefix), and the module ID in the native module.
+    4. Return the module HANDLE, module ID and return code `EBPF_SERVICE_NOT_PRESENT`.
+3. If runtime finds the module, and it not in "unloading" state, it will return a new handle to module.
+4. If runtime finds the module, and it is in "placeholder" state, it will block on an event, which will be set when the module is loaded.
+5. If runtime finds the module, and it is in "unloading" state, it will block on another event, which will be set when call to driver unload returns. Once this event is set, this thread will be unblocked and move to step 2.
+
+#### Misc. Changes
+1. ebpfcore will now maintain 2 hash tables to lookup the native modules. One hash table has service name as key, and another has module ID as key.
+2. Currently native module is deleted from hash table when ref count becomes 0, which is before the driver is unloaded. In this new design the native module should be deleted from the hash table only when the driver unload call has completed. Also, we need to ensure that driver unload happens inline.
+
+### Option 2
+Another approach is to completely skip all the above synchronization logic in ebpfcore and ebpfapi.dll, and separate out service install / uninstall completely from the scope of the application. Native modules will be installed by some installer that installs the native module on the system. Once that is done, programs can be loaded / unloaded multiple times, without the need of install / uninstall of service entries.
+
+Changes needed to implement this:
+1. Installer installs the service entry. We need to ensure the service name is unique. Two native modules with same name but in different locations should be installed as 2 different services. Installer also populates the client module ID registry key with a random GUID.
+2. The app that loads the program will pass the service name to ebpfapi (TBD: How does app know the service name for its service). We can still use the existing `bpf_object__open` API, and the API can take the service name instead of file name as input.
+
+**Note**: If we want to continue using existing APIs without any change, it causes a limitation that with this approach, **every user** of native module will need to implement the installer logic.
```

Comment:
I didn't follow this sentence, can you elaborate?  What is a "user" here?
The native module installer (setup.exe or whatever) needs to implement it.  Calling apps don't, nor do their installers per se.

Reply:
I have updated (re-written) that note. If it still looks incorrect, I can remove that point altogether.